backend/src/Database.hs

```diff
@@ -56,69 +61,108 @@ runDb dbAction = do
+    `leftJoin` table @PlayerStatsYear
       `on` ( \(player :& _ :& yearStats) ->
-               (player ^. PlayerId ==. yearStats ^. PlayerStatsYearPlayerId)
-                 &&. (yearStats ^. PlayerStatsYearYear ==. val year)
+               just (player ^. PlayerId) ==. yearStats ?. PlayerStatsYearPlayerId
+                 &&. (yearStats ?. PlayerStatsYearYear ==. just (val year))
            )
 
-getPlayerStats :: (MonadIO m, MonadLogger m) => PlayerId -> DBAction m (Maybe (Entity PlayerStatsTotal, Entity PlayerStatsYear))
-getPlayerStats pid = do
-  year <- currentYear
-  lift . selectOne $ do
-    (player :& totalStats :& yearStats) <- from $ joinedPlayerStats year
-    where_ (player ^. PlayerId ==. val pid)
-    pure (totalStats, yearStats)
+getPlayerStats :: (MonadIO m, MonadLogger m) => PlayerId -> Year -> DBAction m (Maybe MaybePlayerStats)
+getPlayerStats pid year = lift . selectOne $ do
+  (player :& totalStats :& yearStats) <- from $ joinedPlayerStats year
+  where_ (player ^. PlayerId ==. val pid)
+  pure (totalStats, yearStats)
 
-getAllStats :: (MonadIO m, MonadLogger m) => DBAction m [(Entity Player, Entity PlayerStatsTotal, Entity PlayerStatsYear)]
-getAllStats = do
-  year <- currentYear
+getAllStats :: (MonadIO m, MonadLogger m) => Year -> DBAction m [(Entity Player, MaybePlayerStats)]
+getAllStats year = do
   lift . select $ do
     (player :& totalStats :& yearStats) <- from $ joinedPlayerStats year
-    pure (player, totalStats, yearStats)
-
-replacePlayerStats :: (MonadIO m, MonadLogger m) => PlayerStatsTotal -> PlayerStatsYear -> DBAction m ()
-replacePlayerStats totalStats@(PlayerStatsTotal {..}) yearStats@(PlayerStatsYear {..}) = lift $ do
-  replace (PlayerStatsTotalKey playerStatsTotalPlayerId) totalStats
-  replace (PlayerStatsYearKey playerStatsYearPlayerId playerStatsYearYear) yearStats
-
-insertRatingChange :: (MonadIO m, MonadLogger m) => RatingDiff -> DBAction m ()
-insertRatingChange = lift . insert_
-
-insertGameReport :: (MonadIO m, MonadLogger m) => GameReport -> DBAction m (Key GameReport)
-insertGameReport = lift . insert
-
-getGameReports :: (MonadIO m, MonadLogger m) => DBAction m [(Entity GameReport, Entity Player, Entity Player)]
-getGameReports = lift . select $ do
-  (report :& winner :& loser) <-
-    from $
-      table @GameReport
-        `innerJoin` table @Player
-          `on` (\(report :& winner) -> report ^. GameReportWinnerId ==. winner ^. PlayerId)
-        `innerJoin` table @Player
-          `on` (\(report :& _ :& loser) -> report ^. GameReportLoserId ==. loser ^. PlayerId)
+    pure (player, (totalStats, yearStats))
+
+getInitialStats :: (MonadIO m, MonadLogger m) => DBAction m [Entity PlayerStatsInitial]
+getInitialStats = lift . select $ from $ table @PlayerStatsInitial
+
+joinedGameReports :: From (SqlExpr (Entity GameReport) :& SqlExpr (Entity Player) :& SqlExpr (Entity Player))
+joinedGameReports =
+  table @GameReport
+    `innerJoin` table @Player
```

Comment:
Would love an opinion on if you think an INNER or LEFT join is more appropriate here

Reply:
Had to look up INNER and LEFT haha, it's been a while 😬 My first instinct is LEFT because then any `null` players popping up would let us know there's a data problem + we'd never have game reports go invisibly missing from the display. But I haven't thought through how much complexity that would add since I guess there would need to be handling for the potential `null`s. Plus it's kind of theoretical if we assume the data won't get out of sync, so not sure if it's worth the effort of changing